store/src/java/com/zimbra/cs/service/admin/Auth.java

```diff
@@ -74,378 +56,115 @@
-        Element response = zsc.createElement(AdminConstants.AUTH_RESPONSE);
-        at.encodeAuthResp(response, true);
-
-        /*
-         * bug 67078
-         * also return auth token cookie in http header
-         */
         HttpServletRequest httpReq = (HttpServletRequest)context.get(SoapServlet.SERVLET_REQUEST);
         HttpServletResponse httpResp = (HttpServletResponse)context.get(SoapServlet.SERVLET_RESPONSE);
-        boolean rememberMe = request.getAttributeBool(AdminConstants.A_PERSIST_AUTH_TOKEN_COOKIE, false);
-        at.encode(httpResp, true, ZimbraCookie.secureCookie(httpReq), rememberMe);
 
+        Element response = zsc.createElement(AdminConstants.AUTH_RESPONSE);
         response.addAttribute(AdminConstants.E_LIFETIME, at.getExpires() - System.currentTimeMillis(), Element.Disposition.CONTENT);
-
-        Session session = updateAuthenticatedAccount(zsc, at, context, true);
-        if (session != null) {
-            ZimbraSoapContext.encodeSession(response, session.getSessionId(), session.getSessionType());
+        at.setCsrfTokenEnabled(csrfSupport);
+        if (at.getUsage() == Usage.TWO_FACTOR_AUTH) {
+            response.addUniqueElement(AccountConstants.E_TWO_FACTOR_AUTH_REQUIRED).setText("true");
+        } else {
+            boolean rememberMe = request.getAttributeBool(AdminConstants.A_PERSIST_AUTH_TOKEN_COOKIE, false);
+            at.encode(httpResp, true, ZimbraCookie.secureCookie(httpReq), rememberMe);
+            httpReq.setAttribute(CsrfFilter.AUTH_TOKEN, at);
+            Session session = updateAuthenticatedAccount(zsc, at, context, true);
+            if (session != null) {
+                ZimbraSoapContext.encodeSession(response, session.getSessionId(), session.getSessionType());
+            }
         }
+        at.encodeAuthResp(response, true);
 
-        setCSRFToken(httpReq, httpResp, at, csrfSupport, response);
+        CsrfUtil.setCSRFToken(httpReq, httpResp, at, csrfSupport, response);
 
         return response;
     }
 
-    private void setCSRFToken(HttpServletRequest httpReq, HttpServletResponse httpResp, AuthToken authToken, boolean csrfSupport, Element response) throws ServiceException {
-		boolean csrfCheckEnabled = false;
-		if (httpReq.getAttribute(Provisioning.A_zimbraCsrfTokenCheckEnabled) != null) {
-			csrfCheckEnabled = (Boolean) httpReq.getAttribute(Provisioning.A_zimbraCsrfTokenCheckEnabled);
-		}
-		if (csrfSupport && csrfCheckEnabled) {
-			String accountId = authToken.getAccountId();
-			long authTokenExpiration = authToken.getExpires();
-			int tokenSalt = (Integer) httpReq.getAttribute(CsrfFilter.CSRF_SALT);
-			String token = CsrfUtil.generateCsrfToken(accountId, authTokenExpiration, tokenSalt, authToken);
-			Element csrfResponse = response.addUniqueElement(HeaderConstants.E_CSRFTOKEN);
-			csrfResponse.addText(token);
-			httpResp.setHeader(Constants.CSRF_TOKEN, token);
-		}
-	}
-
```

Comment:
Do not remove this method from here as it will break the existing implementation of the clients. You can use override or overload but do **NOT** remove the existing methods from the Zimbra organisation codebase.

Reply:
This is also a private method, this method is adding the CSRF token in response on the basis of an attribute check. So thought that we can reused this and can move to utility class.